### ISSUE TITLE
Url endpoint document with twisted klein

### ISSUE THREAD

**Issue by NirmalVatsyayan** [twisted/klein#145]
Hi ,
I need a help.
Is there any way to get url end point documentation, in django and flask we could use swagger for the same. Kindly suggest the best possible way to represent apis on UI similar to that, if there is any.

Regards

**Comment by notoriousno** [comment]
There are no swagger/openapi plugins for klein as of this moment, but it wouldn't be too difficult to create one on your own ;) this is a subtle hint that you should make something for the rest of the klein/twisted world to use. I recently made merge request showing how to extend klein #146. Using that example as a reference, it should be possible to create a swagger schema and generate a swagger page. If you need basic endpoint documentation, [Sphinx](http://www.sphinx-doc.org/en/1.4.9/) is very good at creating a simple non-interactive doc.

On a final note, try to ask these kinds of questions on the [Twisted section of Stackoverflow](https://stackoverflow.com/questions/tagged/twisted) in the future.

**Comment by wsanchez** [comment]
Turning this into a feature request

**Comment by SamuelMarks** [comment]
Would be great to have this feature!

FastAPI and others have it; and it's pretty handy.

PS: If there's interest [and I can find the time!] I can build it for you I guess?

**Comment by SamuelMarks** [comment]
Would be great to have this feature!

FastAPI and others have it; and it's pretty handy.

**Comment by SoundsSerious** [comment]
Yes it would be awesome to have klein do this!

I have take a stab at generating a swagger endpoint with return type annotations using a structured gen-ai routine. The assumptions are listed at the top of the file. Good luck!
https://gist.github.com/SoundsSerious/a6bbbddc8f5cfe3b015f038fe25214cd

**Comment by glyph** [comment]
@SoundsSerious I don't think we can accept any genAI contributions due to IP concerns, but I really appreciate you flagging it as such rather than sneaking it into a patch ;)

**Comment by SoundsSerious** [comment]
Thats a good point @glyph, of course anyone could use this gist if they didn't mind going around that kind of restriction. I had been searching for a solution to this and just wanted to share the result because it has been bothering me for years.

I think the core part of this could still work without the gen ai, but you might lose a bit of the magic (and or erronius errors!)

I think the key parts are the route looping, the response formating and then parsing the docstrings which are not genai. 

About the only useful thing the genai was doing was formatting the response, I was hoping it would try to understand the functions a bit more and elaborate but that didn't seem to be effective

**Comment by SoundsSerious** [comment]
If you want to pony up for this feature, I think the solution to this might be to establish a docstring standard for parsing into openapi. The important bits, the path and method, and path-parameters can be determined via klein introspection. The response formatting works with the json-schema.

The description, query arguments and response codes are the only thing that needs to be defined I think

**Comment by SamuelMarks** [comment]
Definitely doable. If enough interest is shown I can finish my project and add support for klein [along with bottle & flask]:
https://github.com/offscale/cdd-python

(Apache-2.0 OR MIT) licensed

**Comment by SamuelMarks** [comment]
Definitely doable. If enough interest is shown I can finish my project and add support for klein [along with bottle & flask]:
https://github.com/SamuelMarks/cdd-python

(Apache-2.0 OR MIT) licensed

**Comment by SoundsSerious** [comment]
@SamuelMarks hey this looks great! Nice job on that library, will definitely be using that for my next project.

Marshmallow might be an interesting one to add as well, that would cover other ORM like ArangoORM.
